### PR TITLE
Remember selected song

### PR DESCRIPTION
**Pull request by davidsbailey**
Fixes https://github.com/code-dot-org/dance-party/issues/375

### CODE REVIEW

**Comment by davidsbailey** [comment]
I've verified that the modified UI test scenario is passing locally.

File: apps/src/dance/Dance.js [hunk 1]
```diff
@@ -171,6 +171,14 @@ Dance.prototype.initSongs = async function (config) {
 
   loadSong(selectedSong, songData);
   this.updateSongMetadata(selectedSong);
+
+  if (config.channel) {
+    // Ensure that the selected song will be stored in the project the first
+    // time we run the level. This ensures that if we are on a project-backed
+    // script level, then the correct song will still be selected after we
+    // share.
+    config.level.selectedSong = selectedSong;
```
Comment (islemaster) [inline apps/src/dance/Dance.js:180]: Nice catch!  Do I understand correctly that `config.level.selectedSong` was set to the level default prior to the first run despite setting the correct selected song into redux, and therefore when we saved the song we were overwriting the correct selected song with the default?

If so, I wonder if we should follow this with a deeper change that deduplicates the "selected song" state that allowed this bug to happen in the first place.  I'd feel better if we treated `config.level` as immutable (I know we mutate it all over the place, but in a perfect world...) and only had one source of truth for which song is selected.

Comment (davidsbailey) [reply apps/src/dance/Dance.js:180]: > Do I understand correctly that config.level.selectedSong was set to the level default prior to the first run despite setting the correct selected song into redux, and therefore when we saved the song we were overwriting the correct selected song with the default?

I'm not sure I understand well enough to answer yes or no to this, but here's what was happening. Previously, we weren't writing any song to main.json, so when you look at the last script level and see Sia, that would never get persisted. So then when you shared your project, you'd have no selectedSong within the project so you'd see the default which was then Katy Perry.

Comment (davidsbailey) [reply apps/src/dance/Dance.js:180]: > If so, I wonder if we should follow this with a deeper change that deduplicates the "selected song" state that allowed this bug to happen in the first place. I'd feel better if we treated config.level as immutable (I know we mutate it all over the place, but in a perfect world...) and only had one source of truth for which song is selected.

This sounds awesome -- the world would be much better if `project.js` could rely on redux state. there might be some way to attempt to do this incrementally. perhaps during HoC lockdown is a good time to try to accomplish this within the staging-next sandbox.

**Pull request merged**